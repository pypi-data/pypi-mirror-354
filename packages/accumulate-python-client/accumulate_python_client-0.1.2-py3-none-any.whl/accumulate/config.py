# accumulate/config.py

# Define Accumulate network base URLs (without "/v3")
# ACCUMULATE_RPC_URL = "https://testnet.accumulatenetwork.io"  # Default: Testnet

ACCUMULATE_RPC_URL = "https://mainnet.accumulatenetwork.io"  # Default: Testnet

# Optional: Add mainnet and custom options
ACCUMULATE_MAINNET_URL = "https://mainnet.accumulatenetwork.io"
ACCUMULATE_TESTNET_URL = "https://testnet.accumulatenetwork.io"
ACCUMULATE_CUSTOM_URL = None  # Set this if using a custom Accumulate node

# Function to get the active Accumulate RPC URL
def get_accumulate_rpc_url():
    """Returns the currently selected Accumulate RPC URL."""
    return ACCUMULATE_CUSTOM_URL or ACCUMULATE_RPC_URL
