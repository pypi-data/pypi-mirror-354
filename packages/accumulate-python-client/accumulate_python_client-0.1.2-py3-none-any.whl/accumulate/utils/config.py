# accumulate-python-client\accumulate\utils\config.py

import os

class Config:
    """
    Configuration utility for managing environment-specific settings.
    """
    @staticmethod
    def is_testnet() -> bool:
        """
        Dynamically get the testnet status based on the environment variable.
        """
        return os.getenv("ACCUMULATE_IS_TESTNET", "false").lower() == "true"

    @staticmethod
    def initial_acme_oracle() -> float:
        """
        Dynamically get the initial ACME oracle value based on the network type.
        """
        return 5000 if Config.is_testnet() else 0.50

    @staticmethod
    def get_network_type() -> str:
        """
        Get the current network type.
        :return: 'testnet' or 'mainnet'
        """
        return "testnet" if Config.is_testnet() else "mainnet"

    @staticmethod
    def get_initial_oracle_value() -> float:
        """
        Get the initial ACME oracle value for the configured network.
        :return: The oracle value as a float.
        """
        return Config.initial_acme_oracle()
