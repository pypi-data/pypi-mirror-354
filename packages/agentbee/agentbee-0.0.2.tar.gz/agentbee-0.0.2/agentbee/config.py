import configparser
from pathlib import Path

CONFIG_DIR = Path.home() / ".agentbee"
CONFIG_FILE = CONFIG_DIR / "config.ini"
CONFIG_SECTION = "DEFAULT"

def get_config_path() -> Path:
    """Returns the path to the config file."""
    return CONFIG_FILE

def save_config(api_key: str, base_url: str, model: str):
    """Saves the configuration to the global config file."""
    CONFIG_DIR.mkdir(exist_ok=True)
    config = configparser.ConfigParser()
    config[CONFIG_SECTION] = {
        'llm_api_key': api_key,
        'llm_base_url': base_url,
        'llm_model': model
    }
    with open(CONFIG_FILE, 'w') as configfile:
        config.write(configfile)
    print(f"âœ… Configuration saved to {CONFIG_FILE}")

def load_config() -> dict:
    if not CONFIG_FILE.exists():
        return {}
    config = configparser.ConfigParser()
    config.read(CONFIG_FILE)
    return {
        'llm_api_key': config.get(CONFIG_SECTION, 'llm_api_key', fallback=''),
        'llm_base_url': config.get(CONFIG_SECTION, 'llm_base_url', fallback=''),
        'llm_model': config.get(CONFIG_SECTION, 'llm_model', fallback='')
    }