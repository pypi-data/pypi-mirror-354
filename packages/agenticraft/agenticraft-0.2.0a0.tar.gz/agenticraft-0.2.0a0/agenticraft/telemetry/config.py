"""Telemetry configuration for AgentiCraft.

This module provides configuration management for OpenTelemetry
settings, including exporters, sampling, and instrumentation options.

Example:
    Basic configuration::

        from agenticraft.telemetry import TelemetryConfig

        config = TelemetryConfig(
            service_name="my-agent-service",
            export_endpoint="http://jaeger:4317",
            sample_rate=0.1,  # Sample 10% of traces
            export_metrics=True,
            export_traces=True
        )
"""

import os
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, model_validator


class ExportFormat(str, Enum):
    """Supported telemetry export formats."""

    OTLP = "otlp"
    JAEGER = "jaeger"
    ZIPKIN = "zipkin"
    CONSOLE = "console"
    NONE = "none"


class TelemetryEnvironment(str, Enum):
    """Deployment environment for telemetry."""

    DEVELOPMENT = "development"
    STAGING = "staging"
    PRODUCTION = "production"
    TEST = "test"


class ExporterConfig(BaseModel):
    """Configuration for telemetry exporters."""

    format: ExportFormat = Field(
        default=ExportFormat.OTLP, description="Export format to use"
    )
    endpoint: str | None = Field(default=None, description="Endpoint for exporter")
    headers: dict[str, str] = Field(
        default_factory=dict, description="Headers to include with exports"
    )
    timeout_ms: int = Field(default=30000, description="Export timeout in milliseconds")
    insecure: bool = Field(
        default=True, description="Whether to use insecure connection"
    )

    @model_validator(mode="after")
    def set_endpoint_defaults(self):
        """Set endpoint defaults based on format."""
        if self.endpoint is None and self.format != ExportFormat.CONSOLE:
            format_map = {
                ExportFormat.OTLP: "http://localhost:4317",
                ExportFormat.JAEGER: "http://localhost:14250",
                ExportFormat.ZIPKIN: "http://localhost:9411/api/v2/spans",
            }
            self.endpoint = format_map.get(self.format)
        return self


class ResourceConfig(BaseModel):
    """Resource attributes for telemetry."""

    service_name: str = Field(default="agenticraft", description="Name of the service")
    service_version: str = Field(
        default="unknown", description="Version of the service"
    )
    service_instance_id: str | None = Field(
        default=None, description="Unique instance identifier"
    )
    environment: TelemetryEnvironment = Field(
        default=TelemetryEnvironment.DEVELOPMENT, description="Deployment environment"
    )
    attributes: dict[str, Any] = Field(
        default_factory=dict, description="Additional resource attributes"
    )


class SamplingConfig(BaseModel):
    """Sampling configuration for traces."""

    sample_rate: float = Field(
        default=1.0, ge=0.0, le=1.0, description="Base sampling rate (0.0-1.0)"
    )
    parent_based: bool = Field(
        default=True, description="Honor parent span sampling decision"
    )
    rules: list[dict[str, Any]] = Field(
        default_factory=list, description="Custom sampling rules"
    )


class InstrumentationConfig(BaseModel):
    """Configuration for automatic instrumentation."""

    instrument_http: bool = Field(default=True, description="Instrument HTTP requests")
    instrument_grpc: bool = Field(default=True, description="Instrument gRPC calls")
    instrument_redis: bool = Field(
        default=False, description="Instrument Redis operations"
    )
    instrument_sqlalchemy: bool = Field(
        default=False, description="Instrument database queries"
    )
    excluded_urls: list[str] = Field(
        default_factory=lambda: ["/health", "/metrics"],
        description="URLs to exclude from tracing",
    )


class TelemetryConfig(BaseModel):
    """Complete telemetry configuration."""

    enabled: bool = Field(default=True, description="Whether telemetry is enabled")

    # Resource configuration
    resource: ResourceConfig = Field(
        default_factory=ResourceConfig, description="Resource configuration"
    )

    # Tracing configuration
    export_traces: bool = Field(default=True, description="Whether to export traces")
    trace_exporter: ExporterConfig = Field(
        default_factory=ExporterConfig, description="Trace exporter configuration"
    )
    sampling: SamplingConfig = Field(
        default_factory=SamplingConfig, description="Trace sampling configuration"
    )

    # Metrics configuration
    export_metrics: bool = Field(default=True, description="Whether to export metrics")
    metric_exporter: ExporterConfig = Field(
        default_factory=ExporterConfig, description="Metric exporter configuration"
    )
    metric_interval_ms: int = Field(
        default=60000, description="Metric export interval in milliseconds"
    )

    # Instrumentation
    instrumentation: InstrumentationConfig = Field(
        default_factory=InstrumentationConfig,
        description="Automatic instrumentation settings",
    )

    # Logging integration
    log_correlation: bool = Field(default=True, description="Add trace IDs to logs")

    @classmethod
    def from_env(cls) -> "TelemetryConfig":
        """Create configuration from environment variables.

        Supports standard OpenTelemetry environment variables:
        - OTEL_SERVICE_NAME
        - OTEL_EXPORTER_OTLP_ENDPOINT
        - OTEL_TRACES_EXPORTER
        - OTEL_METRICS_EXPORTER
        - OTEL_TRACES_SAMPLER_ARG

        Returns:
            Configured TelemetryConfig instance
        """
        config = cls()

        # Service configuration
        if service_name := os.getenv("OTEL_SERVICE_NAME"):
            config.resource.service_name = service_name

        if service_version := os.getenv("OTEL_SERVICE_VERSION"):
            config.resource.service_version = service_version

        # Endpoint configuration
        if endpoint := os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT"):
            config.trace_exporter.endpoint = endpoint
            config.metric_exporter.endpoint = endpoint

        # Trace-specific endpoint
        if trace_endpoint := os.getenv("OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"):
            config.trace_exporter.endpoint = trace_endpoint

        # Metric-specific endpoint
        if metric_endpoint := os.getenv("OTEL_EXPORTER_OTLP_METRICS_ENDPOINT"):
            config.metric_exporter.endpoint = metric_endpoint

        # Headers
        if headers := os.getenv("OTEL_EXPORTER_OTLP_HEADERS"):
            parsed_headers = {}
            for header in headers.split(","):
                if "=" in header:
                    key, value = header.split("=", 1)
                    parsed_headers[key.strip()] = value.strip()
            config.trace_exporter.headers = parsed_headers
            config.metric_exporter.headers = parsed_headers

        # Sampling
        if sample_rate := os.getenv("OTEL_TRACES_SAMPLER_ARG"):
            try:
                config.sampling.sample_rate = float(sample_rate)
            except ValueError:
                pass

        # Exporters
        if traces_exporter := os.getenv("OTEL_TRACES_EXPORTER"):
            if traces_exporter == "none":
                config.export_traces = False
            elif traces_exporter in ExportFormat.__members__.values():
                config.trace_exporter.format = ExportFormat(traces_exporter)

        if metrics_exporter := os.getenv("OTEL_METRICS_EXPORTER"):
            if metrics_exporter == "none":
                config.export_metrics = False
            elif metrics_exporter in ExportFormat.__members__.values():
                config.metric_exporter.format = ExportFormat(metrics_exporter)

        return config

    def to_env_dict(self) -> dict[str, str]:
        """Convert configuration to environment variables.

        Returns:
            Dictionary of environment variables
        """
        env_vars = {
            "OTEL_SERVICE_NAME": self.resource.service_name,
            "OTEL_SERVICE_VERSION": self.resource.service_version,
            "OTEL_TRACES_SAMPLER_ARG": str(self.sampling.sample_rate),
        }

        if self.trace_exporter.endpoint:
            env_vars["OTEL_EXPORTER_OTLP_ENDPOINT"] = self.trace_exporter.endpoint

        if self.trace_exporter.headers:
            headers_str = ",".join(
                f"{k}={v}" for k, v in self.trace_exporter.headers.items()
            )
            env_vars["OTEL_EXPORTER_OTLP_HEADERS"] = headers_str

        if not self.export_traces:
            env_vars["OTEL_TRACES_EXPORTER"] = "none"
        else:
            env_vars["OTEL_TRACES_EXPORTER"] = self.trace_exporter.format.value

        if not self.export_metrics:
            env_vars["OTEL_METRICS_EXPORTER"] = "none"
        else:
            env_vars["OTEL_METRICS_EXPORTER"] = self.metric_exporter.format.value

        return env_vars


# Preset configurations for common scenarios


def development_config() -> TelemetryConfig:
    """Get development environment configuration.

    - Console exporter for easy debugging
    - Full sampling
    - All instrumentation enabled
    """
    return TelemetryConfig(
        resource=ResourceConfig(environment=TelemetryEnvironment.DEVELOPMENT),
        trace_exporter=ExporterConfig(format=ExportFormat.CONSOLE),
        metric_exporter=ExporterConfig(format=ExportFormat.CONSOLE),
        sampling=SamplingConfig(sample_rate=1.0),
    )


def production_config(
    service_name: str, otlp_endpoint: str, sample_rate: float = 0.1
) -> TelemetryConfig:
    """Get production environment configuration.

    Args:
        service_name: Name of your service
        otlp_endpoint: OTLP collector endpoint
        sample_rate: Sampling rate (default 10%)

    Returns:
        Production-ready configuration
    """
    return TelemetryConfig(
        resource=ResourceConfig(
            service_name=service_name, environment=TelemetryEnvironment.PRODUCTION
        ),
        trace_exporter=ExporterConfig(
            format=ExportFormat.OTLP, endpoint=otlp_endpoint, insecure=False
        ),
        metric_exporter=ExporterConfig(
            format=ExportFormat.OTLP, endpoint=otlp_endpoint, insecure=False
        ),
        sampling=SamplingConfig(sample_rate=sample_rate),
    )


def test_config() -> TelemetryConfig:
    """Get test environment configuration.

    - Telemetry disabled by default
    - No external dependencies
    """
    return TelemetryConfig(
        enabled=False, resource=ResourceConfig(environment=TelemetryEnvironment.TEST)
    )
