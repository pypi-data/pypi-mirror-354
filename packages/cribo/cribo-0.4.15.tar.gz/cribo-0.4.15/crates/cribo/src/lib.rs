pub mod ast_rewriter;
pub mod bundler;
pub mod combine;
pub mod config;
pub mod dependency_graph;
pub mod dirs;
pub mod emit;
pub mod resolver;
pub mod unused_import_trimmer;
pub mod unused_imports_simple;
pub mod util;

pub use bundler::Bundler;
pub use config::Config;
