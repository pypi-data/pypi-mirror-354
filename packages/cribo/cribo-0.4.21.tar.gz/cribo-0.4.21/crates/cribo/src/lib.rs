pub mod bundler;
pub mod combine;
pub mod config;
pub mod dependency_graph;
pub mod dirs;
pub mod hybrid_static_bundler;
pub mod resolver;
pub mod unused_imports_simple;
pub mod util;

pub use bundler::Bundler;
pub use config::Config;
