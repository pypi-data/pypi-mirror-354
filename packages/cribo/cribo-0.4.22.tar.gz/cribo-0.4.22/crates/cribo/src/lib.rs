pub mod code_generator;
pub mod combine;
pub mod config;
pub mod dependency_graph;
pub mod dirs;
pub mod orchestrator;
pub mod resolver;
pub mod unused_imports;
pub mod util;

pub use config::Config;
pub use orchestrator::BundleOrchestrator;
