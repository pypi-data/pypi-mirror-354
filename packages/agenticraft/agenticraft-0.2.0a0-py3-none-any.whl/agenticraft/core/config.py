"""Configuration management for AgentiCraft.

This module provides centralized configuration management using
pydantic-settings. Configuration can be loaded from environment
variables, .env files, or set programmatically.

Example:
    Loading configuration::

        from agenticraft.core.config import settings

        # Access configuration
        print(settings.openai_api_key)
        print(settings.default_model)

        # Override configuration
        settings.default_temperature = 0.5
"""

from functools import lru_cache
from pathlib import Path
from typing import Any

from pydantic import Field, field_validator
from pydantic_settings import BaseSettings, SettingsConfigDict


class AgentiCraftSettings(BaseSettings):
    """Global settings for AgentiCraft.

    Settings are loaded from (in order of priority):
    1. Environment variables
    2. .env file in current directory
    3. .env file in parent directories
    4. Default values

    All settings can be overridden using environment variables
    with the AGENTICRAFT_ prefix.
    """

    model_config = SettingsConfigDict(
        env_prefix="AGENTICRAFT_",
        env_file=".env",
        env_file_encoding="utf-8",
        case_sensitive=False,
        extra="allow",
    )

    # General settings
    environment: str = Field(
        default="development",
        description="Environment (development, staging, production)",
    )
    debug: bool = Field(default=False, description="Enable debug mode")
    log_level: str = Field(default="INFO", description="Logging level")

    # API Keys
    openai_api_key: str | None = Field(default=None, description="OpenAI API key")
    anthropic_api_key: str | None = Field(default=None, description="Anthropic API key")
    google_api_key: str | None = Field(default=None, description="Google API key")

    # Model defaults
    default_model: str = Field(default="gpt-4", description="Default LLM model")
    default_temperature: float = Field(
        default=0.7, ge=0.0, le=2.0, description="Default temperature for LLM"
    )
    default_max_tokens: int | None = Field(
        default=None, gt=0, description="Default max tokens for LLM"
    )
    default_timeout: int = Field(
        default=30, gt=0, description="Default timeout in seconds"
    )
    default_max_retries: int = Field(default=3, ge=0, description="Default max retries")

    # Provider settings
    openai_base_url: str = Field(
        default="https://api.openai.com/v1", description="OpenAI API base URL"
    )
    anthropic_base_url: str = Field(
        default="https://api.anthropic.com", description="Anthropic API base URL"
    )
    ollama_base_url: str = Field(
        default="http://localhost:11434", description="Ollama base URL"
    )

    # Memory settings
    memory_backend: str = Field(
        default="sqlite", description="Memory backend (sqlite, json, redis)"
    )
    memory_path: Path = Field(
        default=Path("./memory"), description="Path for memory storage"
    )
    conversation_memory_size: int = Field(
        default=10, gt=0, description="Number of conversation turns to keep"
    )

    # Telemetry settings
    telemetry_enabled: bool = Field(default=True, description="Enable telemetry")
    telemetry_service_name: str = Field(
        default="agenticraft", description="Service name for telemetry"
    )
    telemetry_export_endpoint: str | None = Field(
        default=None, description="OTLP endpoint for telemetry export"
    )
    telemetry_sample_rate: float = Field(
        default=1.0, ge=0.0, le=1.0, description="Telemetry sampling rate"
    )

    # Plugin settings
    plugins_enabled: bool = Field(default=True, description="Enable plugin system")
    plugin_dirs: list[Path] = Field(
        default_factory=lambda: [Path("./plugins")],
        description="Directories to load plugins from",
    )

    # MCP settings
    mcp_enabled: bool = Field(default=True, description="Enable Model Context Protocol")
    mcp_servers: list[str] = Field(
        default_factory=list, description="MCP server URLs to connect to"
    )

    # Tool settings
    tool_execution_timeout: int = Field(
        default=10, gt=0, description="Timeout for tool execution in seconds"
    )
    tool_retry_attempts: int = Field(
        default=2, ge=0, description="Number of retry attempts for failed tools"
    )

    # Workflow settings
    workflow_step_timeout: int = Field(
        default=60, gt=0, description="Default timeout for workflow steps in seconds"
    )
    workflow_max_parallel_steps: int = Field(
        default=5, gt=0, description="Maximum parallel steps in workflows"
    )

    # Security settings
    allow_code_execution: bool = Field(
        default=False, description="Allow execution of arbitrary code in tools"
    )
    allowed_domains: set[str] = Field(
        default_factory=set, description="Allowed domains for web requests"
    )

    @field_validator("log_level")
    def validate_log_level(cls, v: str) -> str:
        """Validate log level."""
        valid_levels = {"DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"}
        v_upper = v.upper()
        if v_upper not in valid_levels:
            raise ValueError(f"Invalid log level: {v}. Must be one of {valid_levels}")
        return v_upper

    @field_validator("environment")
    def validate_environment(cls, v: str) -> str:
        """Validate environment."""
        valid_envs = {"development", "staging", "production", "test"}
        v_lower = v.lower()
        if v_lower not in valid_envs:
            raise ValueError(f"Invalid environment: {v}. Must be one of {valid_envs}")
        return v_lower

    @field_validator("memory_backend")
    def validate_memory_backend(cls, v: str) -> str:
        """Validate memory backend."""
        valid_backends = {"sqlite", "json", "redis"}
        v_lower = v.lower()
        if v_lower not in valid_backends:
            raise ValueError(
                f"Invalid memory backend: {v}. Must be one of {valid_backends}"
            )
        return v_lower

    def get_api_key(self, provider: str) -> str | None:
        """Get API key for a provider.

        Args:
            provider: Provider name (openai, anthropic, google)

        Returns:
            API key if available
        """
        key_map = {
            "openai": self.openai_api_key,
            "anthropic": self.anthropic_api_key,
            "google": self.google_api_key,
        }
        return key_map.get(provider.lower())

    def get_base_url(self, provider: str) -> str | None:
        """Get base URL for a provider.

        Args:
            provider: Provider name

        Returns:
            Base URL if available
        """
        url_map = {
            "openai": self.openai_base_url,
            "anthropic": self.anthropic_base_url,
            "ollama": self.ollama_base_url,
        }
        return url_map.get(provider.lower())

    def to_agent_config(self) -> dict[str, Any]:
        """Convert settings to agent configuration.

        Returns:
            Dictionary suitable for Agent initialization
        """
        return {
            "model": self.default_model,
            "temperature": self.default_temperature,
            "max_tokens": self.default_max_tokens,
            "timeout": self.default_timeout,
            "max_retries": self.default_max_retries,
        }

    def to_telemetry_config(self) -> dict[str, Any]:
        """Convert settings to telemetry configuration.

        Returns:
            Dictionary suitable for Telemetry initialization
        """
        return {
            "service_name": self.telemetry_service_name,
            "export_to": self.telemetry_export_endpoint,
            "enabled": self.telemetry_enabled,
            "sample_rate": self.telemetry_sample_rate,
        }

    @property
    def is_production(self) -> bool:
        """Check if running in production."""
        return self.environment == "production"

    @property
    def is_development(self) -> bool:
        """Check if running in development."""
        return self.environment == "development"

    def validate_required_keys(self, providers: list[str]) -> None:
        """Validate that required API keys are present.

        Args:
            providers: List of providers that need keys

        Raises:
            ValueError: If required keys are missing
        """
        missing = []
        for provider in providers:
            if not self.get_api_key(provider):
                missing.append(provider)

        if missing:
            raise ValueError(
                f"Missing API keys for providers: {', '.join(missing)}. "
                f"Set environment variables or update .env file."
            )


@lru_cache
def get_settings() -> AgentiCraftSettings:
    """Get the global settings instance (cached).

    Returns:
        AgentiCraftSettings instance
    """
    return AgentiCraftSettings()


# Global settings instance
settings = get_settings()


def reload_settings() -> AgentiCraftSettings:
    """Reload settings from environment.

    This clears the cache and creates a new settings instance.
    Useful for testing or when environment variables change.

    Returns:
        New AgentiCraftSettings instance
    """
    get_settings.cache_clear()
    global settings
    settings = get_settings()
    return settings


def update_settings(**kwargs: Any) -> None:
    """Update settings programmatically.

    Args:
        **kwargs: Setting values to update

    Example:
        update_settings(default_model="gpt-3.5-turbo", debug=True)
    """
    for key, value in kwargs.items():
        if hasattr(settings, key):
            setattr(settings, key, value)
        else:
            raise ValueError(f"Unknown setting: {key}")
