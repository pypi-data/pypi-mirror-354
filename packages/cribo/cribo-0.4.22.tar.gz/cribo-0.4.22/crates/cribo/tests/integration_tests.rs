#![allow(clippy::disallowed_methods)]

use insta::{assert_snapshot, with_settings};
use std::path::PathBuf;
use tempfile::TempDir;

use cribo::config::Config;
use cribo::orchestrator::BundleOrchestrator;

#[test]
fn test_simple_project_bundling() {
    // Initialize logger for debugging
    let _ = env_logger::try_init();

    let config = Config {
        src: vec![PathBuf::from("tests/fixtures/simple_project")],
        ..Default::default()
    };
    let mut bundler = BundleOrchestrator::new(config);

    // Create temporary output directory
    let temp_dir = TempDir::new().unwrap();
    let output_path = temp_dir.path().join("bundle.py");

    // Path to our test fixture
    let entry_path = PathBuf::from("tests/fixtures/simple_project/main.py");

    // Skip if test fixture doesn't exist (for CI environments)
    if !entry_path.exists() {
        return;
    }

    // Test bundling
    let result = bundler.bundle(&entry_path, &output_path, false);

    match result {
        Ok(()) => {
            // Verify output file exists
            assert!(output_path.exists());

            // Read and snapshot the bundled content
            let content = std::fs::read_to_string(&output_path).unwrap();

            with_settings!({
                description => "Bundle output for simple project with User model and helper functions",
            }, {
                assert_snapshot!(content);
            });
        }
        Err(e) => {
            eprintln!("Bundle test failed: {}", e);
            // Print more error details
            println!("Full error chain:");
            let mut source = e.source();
            while let Some(err) = source {
                println!("  Caused by: {}", err);
                source = err.source();
            }
            panic!("Bundling failed: {}", e);
        }
    }
}

#[test]
fn test_requirements_generation() {
    let config = Config::default();
    let mut bundler = BundleOrchestrator::new(config);

    // Create temporary output directory
    let temp_dir = TempDir::new().unwrap();
    let output_path = temp_dir.path().join("bundle.py");

    // Path to our test fixture with third-party dependencies
    let entry_path = PathBuf::from("tests/fixtures/pydantic_project/main.py");

    // Skip if test fixture doesn't exist
    if !entry_path.exists() {
        return;
    }

    // Test bundling with requirements generation
    let result = bundler.bundle(&entry_path, &output_path, true);

    match result {
        Ok(()) => {
            let mut output = String::new();

            // Check if bundle was created
            if output_path.exists() {
                let bundle_content = std::fs::read_to_string(&output_path).unwrap();
                output.push_str("=== BUNDLE CONTENT ===\n");
                output.push_str(&bundle_content);
                output.push_str("\n=== END BUNDLE CONTENT ===\n\n");
            }

            // Check if requirements.txt was created
            let requirements_path = temp_dir.path().join("requirements.txt");
            if requirements_path.exists() {
                let requirements = std::fs::read_to_string(&requirements_path).unwrap();
                output.push_str("=== REQUIREMENTS.TXT ===\n");
                output.push_str(&requirements);
                output.push_str("\n=== END REQUIREMENTS.TXT ===");
            } else {
                output.push_str("No requirements.txt file was generated.");
            }

            with_settings!({
                description => "Bundle and requirements generation for project with third-party dependencies",
            }, {
                assert_snapshot!(output);
            });
        }
        Err(e) => {
            eprintln!("Requirements test failed: {}", e);
            panic!("Requirements generation test failed: {}", e);
        }
    }
}

#[test]
fn test_module_resolution() {
    use cribo::config::Config;
    use cribo::resolver::{ImportType, ModuleResolver};

    let config = Config {
        src: vec![PathBuf::from("tests/fixtures/simple_project")],
        ..Default::default()
    };

    let resolver = ModuleResolver::new(config);

    match resolver {
        Ok(resolver) => {
            let first_party_modules_set = resolver.get_first_party_modules();
            let mut first_party_modules: Vec<_> = first_party_modules_set.iter().cloned().collect();
            first_party_modules.sort(); // Sort for deterministic output

            // Test import classifications
            let classifications = [
                (
                    ".relative_module",
                    resolver.classify_import(".relative_module"),
                ),
                (
                    "..parent_module",
                    resolver.classify_import("..parent_module"),
                ),
                (
                    "...grandparent.module",
                    resolver.classify_import("...grandparent.module"),
                ),
                ("os", resolver.classify_import("os")),
                (
                    "unknown_package",
                    resolver.classify_import("unknown_package"),
                ),
            ];

            // Verify the expected classifications
            assert_eq!(classifications[0].1, ImportType::FirstParty);
            assert_eq!(classifications[1].1, ImportType::FirstParty);
            assert_eq!(classifications[2].1, ImportType::FirstParty);
            assert_eq!(classifications[3].1, ImportType::StandardLibrary);
            assert_eq!(classifications[4].1, ImportType::ThirdParty);

            let output = format!(
                "First-party modules discovered: {:#?}\n\nImport classifications:\n{}",
                first_party_modules, // Use the sorted Vec
                classifications
                    .iter()
                    .map(|(module, import_type)| format!("  {} -> {:?}", module, import_type))
                    .collect::<Vec<_>>()
                    .join("\n")
            );

            with_settings!({
                description => "Module resolution results showing discovered modules and import classifications",
            }, {
                assert_snapshot!(output);
            });
        }
        Err(e) => {
            eprintln!("Module resolution test failed: {}", e);
            panic!("Module resolution failed: {}", e);
        }
    }
}

#[test]
fn test_dependency_graph() {
    use cribo::dependency_graph::{DependencyGraph, ModuleNode};
    use std::path::PathBuf;

    let mut graph = DependencyGraph::new();

    // Create test modules
    let main_module = ModuleNode {
        name: "main".to_string(),
        path: PathBuf::from("main.py"),
        imports: vec!["utils.helpers".to_string(), "models.user".to_string()],
    };

    let utils_module = ModuleNode {
        name: "utils.helpers".to_string(),
        path: PathBuf::from("utils/helpers.py"),
        imports: vec![],
    };

    let models_module = ModuleNode {
        name: "models.user".to_string(),
        path: PathBuf::from("models/user.py"),
        imports: vec![],
    };

    // Add modules to graph
    graph.add_module(main_module);
    graph.add_module(utils_module);
    graph.add_module(models_module);

    // Add dependencies - main depends on utils and models
    // This means utils.helpers -> main and models.user -> main (dependency -> dependent)
    graph.add_dependency("utils.helpers", "main").unwrap();
    graph.add_dependency("models.user", "main").unwrap();

    // Collect graph information
    let mut modules = graph.get_modules();
    modules.sort_by(|a, b| a.name.cmp(&b.name));
    let module_count = modules.len();
    let mut dependencies_info = Vec::new();

    for module in modules {
        if let Some(deps) = graph.get_dependencies(&module.name) {
            dependencies_info.push(format!("Module {} depends on: {:?}", module.name, deps));
        }
    }

    // Test topological sort
    let sorted = graph.topological_sort().unwrap();
    let sorted_names: Vec<&str> = sorted.iter().map(|m| m.name.as_str()).collect();

    let output = format!(
        "Graph has {} modules\n\nDependencies:\n{}\n\nTopological sort: {:?}",
        module_count,
        dependencies_info.join("\n"),
        sorted_names
    );

    with_settings!({
        description => "Dependency graph creation and topological sorting results",
    }, {
        assert_snapshot!(output);
    });

    // Verify the sort order still makes sense
    let main_index = sorted.iter().position(|m| m.name == "main").unwrap();
    let utils_index = sorted
        .iter()
        .position(|m| m.name == "utils.helpers")
        .unwrap();
    let models_index = sorted.iter().position(|m| m.name == "models.user").unwrap();

    assert!(utils_index < main_index);
    assert!(models_index < main_index);
}

#[test]
fn test_extract_edge_case_imports() {
    // Initialize logger for debugging
    let _ = env_logger::try_init();

    let bundler = BundleOrchestrator::new(Config::default());
    // Path to edge-case test file in the fixtures directory
    let file_path = PathBuf::from("tests/fixtures/test_edge_cases.py");

    // Ensure the test file exists
    assert!(
        file_path.exists(),
        "test_edge_cases.py not found at expected location"
    );

    let imports = bundler
        .extract_imports(&file_path, None)
        .expect("Failed to extract imports");

    let output = format!(
        "Extracted imports from edge cases file:\n{}",
        imports
            .iter()
            .enumerate()
            .map(|(i, import)| format!("  {}: {}", i + 1, import))
            .collect::<Vec<_>>()
            .join("\n")
    );

    with_settings!({
        description => "Import extraction from complex Python file with various import patterns",
    }, {
        assert_snapshot!(output);
    });
}
