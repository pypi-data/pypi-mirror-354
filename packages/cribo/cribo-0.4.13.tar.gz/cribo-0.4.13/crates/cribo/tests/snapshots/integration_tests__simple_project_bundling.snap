---
source: crates/cribo/tests/integration_tests.rs
description: Bundle output for simple project with User model and helper functions
expression: content
---
#!/usr/bin/env python3
# Generated by Cribo - Python Source Bundler
# https://github.com/ophidiarium/cribo

# Preserved imports
import os
import sys

# ─ Module: utils.helpers ─
import types
utils = types.ModuleType('utils')
utils.helpers = types.ModuleType('utils.helpers')
exec('"""Utility functions for the test project."""\nimport os\nimport sys\nfrom typing import Union\ndef greet(name: str) -> str:\n    """Greet a person by name."""\n    return f"Hello, {name}!"\ndef calculate(a: Union[int, float], b: Union[int, float]) -> Union[int, float]:\n    """Calculate the sum of two numbers."""\n    return a + b\ndef get_system_info() -> dict:\n    """Get basic system information."""\n    return {"platform": sys.platform, "python_version": sys.version, "cwd": os.getcwd()}', utils.helpers.__dict__)

# ─ Module: models.user ─
import types
models = types.ModuleType('models')
models.user = types.ModuleType('models.user')
exec('"""User model for the test project."""\nfrom dataclasses import dataclass\nfrom typing import Optional\n@dataclass\nclass User:\n    """A simple user model."""\n    name: str\n    age: int\n    email: Optional[str] = None\n\n    def is_adult(self) -> bool:\n        """Check if the user is an adult."""\n        return self.age >= 18\n\n    def get_display_name(self) -> str:\n        """Get a display name for the user."""\n        if self.email:\n            return f"{self.name} <{self.email}>"\n        return self.name', models.user.__dict__)

# ─ Module: utils ─
greet = utils.helpers.greet
calculate = utils.helpers.calculate
get_system_info = utils.helpers.get_system_info
"""Utils package initialization."""
__all__ = ["greet", "calculate", "get_system_info"]

# ─ Module: models ─
User = models.user.User
"""Models package initialization."""
__all__ = ["User"]

# ─ Entry Module: main ─
greet = utils.helpers.greet
calculate = utils.helpers.calculate
User = models.user.User
"""Simple test project entry point."""
def main():
    """Main function."""
    user = User("Alice", 30)
    print(greet(user.name))
    result = calculate(10, 20)
    print(f"Calculation result: {result}")
    print(f"User: {user.name}, Age: {user.age}")
if __name__ == "__main__":
    main()
